src/Microsoft.AspNet.Mvc.Core/ModelBinding/GenericModelBinder.cs

```diff
@@ -39,57 +68,65 @@ private static Type ResolveBinderType(Type modelType)
                    GetKeyValuePairBinder(modelType);
         }
 
-        private static Type GetArrayBinder(Type modelType)
+        private static GenericModelBindingInfo GetArrayBinder(Type modelType)
         {
             if (modelType.IsArray)
             {
                 var elementType = modelType.GetElementType();
-                return typeof(ArrayModelBinder<>).MakeGenericType(elementType);
+                var modelBinderType = typeof(ArrayModelBinder<>).MakeGenericType(elementType);
+
+                return new GenericModelBindingInfo()
+                {
+                    ModelBinderType = modelBinderType,
+                    UnderlyingModelType = modelType,
+                };
             }
+
             return null;
         }
 
-        private static Type GetCollectionBinder(Type modelType)
+        private static GenericModelBindingInfo GetCollectionBinder(Type modelType)
```

Comment:
Suggest renaming this method and similar ones below.
GetCollectionBinder => GetCollectionBinderInfo